File: object_oriented.cpp
//create TXT files as array1.txt and array2.txt which stores values of the array

#include <bits/stdc++.h>
using namespace std;

class Object_oriented
{
private:
    //creating private variables
    int array1[100];
    int array2[100];
    int size1;
    int size2;

    //creating unordered map
    unordered_map<int, int> umap;

public:
    //constructor 
    Object_oriented() {}

    //Method to find the soulutio i.e operations on the array
    void findSolution()
    {
        //for loop which itereates through the array1 to initialize each unique key to 0
        for (int i = 0; i < size1; i++)
        {
            umap[array1[i]] = 0;
        }
        //for loop to check for the presence of the target element from array2 and validate in the map we created
        for (int i = 0; i < size2; i++)
        {
            if (umap.find(array2[i]) != umap.end())
            {
                umap[array2[i]]++; //incrementing the key's value
            }
        }
    }

    //Method to print the result
    void printResult()
    {   
        //printing the output by iteratin to each elements from the map we created
        //here the auto i is an iterator which reads the each positional values in the unordermap
        for (auto i = umap.begin(); i != umap.end(); ++i)
        {
            cout << i->first << " : " << i->second << endl;
        }
    }

    //Method to take input for array1 and array2
    void takeInput()
    {
        //taking input from a file in which each line contains element of array1
        ifstream file1; //we declare a variable to store the file data
        size1 = 0; //initializing the size1 variable to 0 so that we can track array size parallelly
        file1.open("array1.txt"); //opening the array1.txt which has the values of the array
        while (file1 >> array1[size1]) //pushing values into the respective positions inside array
        {
            size1++;
        }

        //taking input from a file in which each line contains element of array2 
        //rest same as above
        ifstream file2;
        size2 = 0;
        file2.open("array2.txt");
        while (file2 >> array2[size2])
        {
            size2++;
        }
        

        // cout << "Enter the size of input for array1 ";
        // cin >> size1;
        // cout << "Enter the elements for array1: ";
        // for (int i = 0; i < size1; i++)
        // {
        //     cin >> array1[i];
        // }
        // cout << "Enter the size of input for array2 ";
        // cin >> size2;
        // cout << "Enter the elements for array2: ";
        // for (int i = 0; i < size2; i++)
        // {
        //     cin >> array2[i];
        // }
    }
    
};

int main()
{
    //variable to store the current time and end time
    time_t start_time, end_time;


    //creating object of class
    Object_oriented nobject;

    //creating object for taking input
    nobject.takeInput();

    //store the current time
    time(&start_time);

    //creating object for finding solution
    nobject.findSolution();

    //creating object for printing result
    nobject.printResult();

    //store the end time
    time(&end_time);

    //calculate the total time taken
    double time_taken = double(end_time - start_time);

    //print the total time taken
    cout << "The total time taken: " << fixed << time_taken << setprecision(5);
    // cout << "The total time taken: " << time_taken << " seconds" << endl;


    return 0;
}

File: go/object_oriented.go
package main

import (
    "fmt"
    "os"
    "bufio"
    "time"
)

type ObjectOriented struct {
    array1 []int
    array2 []int
    umap   map[int]int
}

func NewObjectOriented() *ObjectOriented {
    return &ObjectOriented{
        umap: make(map[int]int),
    }
}

func (o *ObjectOriented) findSolution() {
    for _, val := range o.array1 {
        o.umap[val] = 0
    }
    for _, val := range o.array2 {
        if _, ok := o.umap[val]; ok {
            o.umap[val]++
        }
    }
}

func (o *ObjectOriented) printResult() {
    for key, value := range o.umap {
        fmt.Printf("%d : %d\n", key, value)
    }
}

func (o *ObjectOriented) takeInput() {
    file1, err := os.Open("array1.txt")
    if err != nil {
        fmt.Println("Error opening array1.txt:", err)
        os.Exit(1)
    }
    defer file1.Close()

    scanner1 := bufio.NewScanner(file1)
    for scanner1.Scan() {
        val := 0
        fmt.Sscanf(scanner1.Text(), "%d", &val)
        o.array1 = append(o.array1, val)
    }

    file2, err := os.Open("array2.txt")
    if err != nil {
        fmt.Println("Error opening array2.txt:", err)
        os.Exit(1)
    }
    defer file2.Close()

    scanner2 := bufio.NewScanner(file2)
    for scanner2.Scan() {
        val := 0
        fmt.Sscanf(scanner2.Text(), "%d", &val)
        o.array2 = append(o.array2, val)
    }
}

func main() {
    start := time.Now()

    object := NewObjectOriented()
    object.takeInput()
    object.findSolution()
    object.printResult()

    elapsed := time.Since(start)
    fmt.Printf("The total time taken: %v\n", elapsed)
}
// array1.txt array2.txt